### PR TITLE
Particle Analyser: add branch length and branch count measurements

### PR DESCRIPTION
**Pull request by mdoube**
Runs Skeletonize3D and AnalyzeSkeleton on each particle individually.

This is in response to #214 and https://forum.image.sc/t/counting-number-of-pores-in-3d/34092/13

### CODE REVIEW

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java [hunk 1]
```diff
@@ -516,6 +520,45 @@ public void filterParticles(final ImagePlus imp, final int[][] particleLabels, f
...
+	 * 
+	 * Uses AnalyzeSkeleton's run() method with no arguments.
+	 * 
+	 * @param imp Input image
+	 * @param particleLabels particle label array
+	 * @param limits xyz limits of each particle
+	 * @param nParticles number of particles
+	 * @return array of AnalyzeSkelton result objects.
+	 * 	Use methods therein to get specific results out. Each result is a list of trees,
+	 * but there should be only one tree per particle so its result is at [0].
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: @return array of {@link SkeletonResult} for each particle

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java [hunk 2]
```diff
@@ -516,6 +520,45 @@ public void filterParticles(final ImagePlus imp, final int[][] particleLabels, f
 		}
 		return eulerCharacters;
 	}
+	
+	/**
+	 * Calculate number of branches and total branch length by running 
+	 * Skeletonize3D and Analyze Skeleton on each particle.
+	 * 
+	 * Uses AnalyzeSkeleton's run() method with no arguments.
+	 *
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]:
```
<p>
Uses AnalyzeSkeleton's run() method with no arguments.
</p>
```

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java [hunk 3]
```diff
@@ -516,6 +520,45 @@ public void filterParticles(final ImagePlus imp, final int[][] particleLabels, f
...
+	 * 	Use methods therein to get specific results out. Each result is a list of trees,
+	 * but there should be only one tree per particle so its result is at [0].
+	 */
+	static SkeletonResult[] getBranchLength(final ImagePlus imp,
+			final int[][] particleLabels, final int[][] limits, final int nParticles)
+	{
+		final SkeletonResult[] skeletonResults = new SkeletonResult[nParticles];
+		Thread[] threads = Multithreader.newThreads();
+		AtomicInteger ai = new AtomicInteger(1);
+		for (int thread = 0; thread < threads.length; thread++) {
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: Are `AnalyzeSkeleton_` and `Skeletonize` thread safe?

File: Legacy/bonej/pom.xml [hunk 4]
```diff
@@ -116,13 +116,21 @@
             <groupId>sc.fiji</groupId>
             <artifactId>3D_Viewer</artifactId>
         </dependency>
+        <dependency>
+            <groupId>sc.fiji</groupId>
+            <artifactId>VIB_</artifactId>
+        </dependency>
```
Comment (rimadoma) [inline Legacy/bonej/pom.xml]: Check if VIB_ is still needed with `mvn dependency:analyze`

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java [hunk 5]
```diff
@@ -516,6 +520,45 @@ public void filterParticles(final ImagePlus imp, final int[][] particleLabels, f
...
+	 * @return array of AnalyzeSkelton result objects.
+	 * 	Use methods therein to get specific results out. Each result is a list of trees,
+	 * but there should be only one tree per particle so its result is at [0].
+	 */
+	static SkeletonResult[] getBranchLength(final ImagePlus imp,
+			final int[][] particleLabels, final int[][] limits, final int nParticles)
+	{
+		final SkeletonResult[] skeletonResults = new SkeletonResult[nParticles];
+		Thread[] threads = Multithreader.newThreads();
+		AtomicInteger ai = new AtomicInteger(1);
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: These can be `final`

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java [hunk 6]
```diff
@@ -516,6 +520,45 @@ public void filterParticles(final ImagePlus imp, final int[][] particleLabels, f
...
+		for (int thread = 0; thread < threads.length; thread++) {
+			threads[thread] = new Thread(() -> {
+				for (int i = ai.getAndIncrement(); i < nParticles; i = ai.getAndIncrement()) {
+					final ImagePlus particleImp = getBinaryParticle(i, imp, particleLabels,
+							limits, 1);
+					final PlugInFilter skeletoniser = new Skeletonize3D_();
+					skeletoniser.setup("", particleImp);
+					skeletoniser.run(null);
+					final AnalyzeSkeleton_ analyzeSkeleton_ = new AnalyzeSkeleton_();
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: Each thread can create their `AnalyzeSkeleton_` and `Skeletonize_` instances once outside the loop

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java [hunk 7]
```diff
@@ -106,8 +107,8 @@ public void run(final String arg) {
 		final String units = cal.getUnits();
 		final GenericDialog gd = new GenericDialog("Setup");
 		final String[] headers = { "Measurement Options", " " };
-		final String[] labels = new String[10];
-		final boolean[] defaultValues = new boolean[10];
+		final String[] labels = new String[12];
+		final boolean[] defaultValues = new boolean[12];
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java]: I think these would be clearer as
```
final String[] labels = {"Exclude on sides", ... "Skeletons"};
final boolean[] defaultValues = {false, ... false};
``` 
If most defaults are false you could do
```
final boolean[] defaultValues = new boolean[labels.length];
defaultValues[8] = true; 
```

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java [hunk 8]
```diff
@@ -262,6 +266,10 @@ public void run(final String arg) {
 		if (doEllipsoids || doEllipsoidImage || doEllipsoidStack) {
 			ellipsoids = ParticleAnalysis.getEllipsoids(surfacePoints);
 		}
+		SkeletonResult[] skeletonResults = new SkeletonResult[nParticles];
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java]: `getBranchLength` creates an instance, so `SkeletonResult[] skeletonResult = null`

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java [hunk 9]
```diff
@@ -300,6 +308,20 @@ public void run(final String arg) {
 						rt.addValue("vZ2", E.getV().get(2, 2));
 					}
 				}
+				if (doSkeletons) {
```
Comment (rimadoma) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java]: `if (skeletonResult) != null`?

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java (hunk 2 above)
Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: Neat trick.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java (hunk 3 above)
Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: No idea, will test.

Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: (nothing spewed out on me so far, but I will try again with more faster threads)

File: Legacy/bonej/pom.xml (hunk 4 above)
Comment (mdoube) [reply Legacy/bonej/pom.xml]: You are right:
```
[WARNING] Unused declared dependencies found:
[WARNING]    sc.fiji:VIB_:jar:3.0.2:compile
```
also:
```
[WARNING] Unused declared dependencies found:
[WARNING]    org.scijava:vecmath:jar:1.6.0-scijava-2:compile
```
and:
```
[WARNING] Unused declared dependencies found:
[WARNING]    com.miglayout:miglayout-swing:jar:5.2:compile
```

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java (hunk 6 above)
Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: Done.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java (hunk 7 above)
Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java]: It's easier to add and remove items if the array index is next to the String: harder to lose count of which label belongs to which value.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java (hunk 5 above)
Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleAnalysis.java]: OK

File: Legacy/bonej/pom.xml (hunk 4 above)
Comment (mdoube) [reply Legacy/bonej/pom.xml]: that last one about `miglayout` may require separate consideration.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java (hunk 8 above)
Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java]: OK

File: Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java (hunk 9 above)
Comment (mdoube) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java]: Prefer this way, it's clearer communication of intent. `skeletonResult` is only not `null` if `doSkeletons = true`

File: Legacy/bonej/pom.xml (hunk 4 above)
Comment (rimadoma) [reply Legacy/bonej/pom.xml]: Miglayout IS needed runtime or GUI borks

**Pull request merged**